### ISSUE TITLE
Proposal: Do not (by default) squash commits when merging

### ISSUE THREAD

**Issue by maximiliankolb** [theforeman/foreman-documentation#884]
Hi all,
after digging around in git log to find where, how, why, and who changed things, I -again- tripped over the default setting to squash commits when merging using the "Squash and merge" button in Github.

Some thoughts:
* One change should be one commit: for example, changing the links to a procedure in multiple files is one commit. Fixing a typo in the same file should be a separate commit.
* (Probably for CLI users only) You can use `!fixup` commits to indicate squashing when merging, for example using `git commit --fixup=head` to create a fixup commit for the last commit; or `git commit --fixup=MY_SHA1` to create a fixup commit for a specific commit. Note that you must only do fixup commits for commits that have not been merged yet, i.e. only for commits that are exclusively on your current branch. Run `git rebase -i master` to rebase your branch to master before pushing; git will automatically correct the order of "fixup commits" and squash (technically: fixup^^) them).
* If you are using software (read: not `git` in your console) that uses the same commit message as your first commit on your branch, then yes, it absolutely makes sense to squash those commits before merging; we do not want multiple commits like "Fix broken xref" merged to master.

I see a lot of advantages when using `git diff`, `git log`, and `git blame`. This also makes cherry-picking and reverting much easier and safer.

For me, it is no accident when I create a PR with more than one commit.

I would also like to enable "merge commits", but I fear this might be pushing it^^.

What are your thoughts? Does anyone else experience the same issues?

**Comment by melcorr** [comment]
@lzap what do you think?

**Comment by ekohl** [comment]
In my experience merging things (and caring about a good git history) the answer is (as always) "it depends".

When there is a single commit, I always use Rebase and merge if the commit message is alright. What's right, again, depends. In this project I'm less strict in what others do but I like to hold myself to high standards. That means describing what you did and, more importantly, why.

Sometimes the commit message is not great. If it's a single commit, I sometimes fix it up using Squash and merge. Which brings me to the next topic. I dislike the "Squash and merge" message that GH generates. I never like it as-is and if I do use the functionality, I always edit it to make sense. This also means stripping the PR number from the commit message. I also think the GH dialog doesn't write hard line endings which is also against the convention so I often write it in vim, format it and copy it back (back in the day I had a Firefox extension for this).

I also use Squash and merge if there's obvious fixups needed in the commits.

> For me, it is no accident when I create a PR with more than one commit.

This is also true for me. You can generally speaking recognize if the author intended multiple commits. There is generally more care in what each commit does: they are atomic (only change one thing) and commit messages are more descriptive.

This should be respected. Either by merging as rebases or with merge commits.

That brings me to my last point: I dislike a git history with a lot of merge commits. In particular with GH you end up with a lot of "one commit"-branches that are then merged straight away but with a merge commit. I like git's default merging better: only create a merge commit if it's not a fast forward merge.

I hope this helps. They are my thoughts in general about git, GH and merging. They probably apply here as well but there are always exceptions to rules.

**Comment by lzap** [comment]
The only reason why I kept this turned on is that many writers would file PR with many WIP commits and I the easiest approach for me is to simply squash them into meaningful commits. If I used this accidentally on a PR that was intentionally filed with multiple commits, then my apologies.

I do not have a strong opinion, we can turn it off and require authors to squash prior merging, I am fine with that. Or we (I) can also be more cautious when using this feature. Whatever you guys prefer.
